File: backend/app/schemas/agent.py
"""Request/response schemas for the Agent CRUD API."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class AgentCreate(BaseModel):
    """Fields required to create a new agent."""

    name: str
    role: str

    engine_type: str
    engine_provider: str
    engine_model: str | None = None
    engine_command: str | None = None
    working_directory: str | None = None
    allowed_tools: list[str] | None = None
    # Plaintext in the request only — encrypted before storage (see
    # api/agents.py's create_agent) and never returned; AgentRead exposes
    # only has_engine_api_key. Required for a bring-your-own API provider
    # (engine_provider outside {"ollama", "anthropic"}), validated in the route.
    engine_api_key: str | None = None

    personality_traits: list[str]
    personality_quirks: list[str] | None = None
    voice_language: str = "en"
    voice_gender: str = "female"
    voice_pace: str = "medium"

    character_id: str | None = None
    accent_color: str = "#6366f1"
    desk_position: int | None = None


class AgentUpdate(BaseModel):
    """Fields that may be patched on an existing agent. All optional."""

    name: str | None = None
    role: str | None = None

    engine_type: str | None = None
    engine_provider: str | None = None
    engine_model: str | None = None
    engine_command: str | None = None
    working_directory: str | None = None
    allowed_tools: list[str] | None = None
    # Same plaintext-in/encrypted-at-rest contract as AgentCreate. Sent as
    # "" to explicitly clear a stored key; omitted entirely (the default,
    # via exclude_unset) to leave whatever's already stored untouched.
    engine_api_key: str | None = None

    personality_traits: list[str] | None = None
    personality_quirks: list[str] | None = None
    voice_language: str | None = None
    voice_gender: str | None = None
    voice_pace: str | None = None

    character_id: str | None = None
    accent_color: str | None = None
    desk_position: int | None = None

    status: str | None = None
    mood: str | None = None


class AgentRead(BaseModel):
    """Full agent representation returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    role: str

    engine_type: str
    engine_provider: str
    engine_model: str | None
    engine_command: str | None
    working_directory: str | None
    allowed_tools: list[str] | None
    # Never the raw key — just whether one is configured (see
    # Agent.has_engine_api_key in models/agent.py).
    has_engine_api_key: bool
    # Whether this agent is anyone's teammate (agent_collaborators) — not a
    # real column, set as a transient attribute by api/agents.py's
    # _mark_sub_agents() before serialization. Drives the frontend's "only
    # chat with main agents" rule. Defaults False defensively (rather than
    # a required field) so a route that forgets to call _mark_sub_agents
    # degrades to "looks like a main agent" instead of a 500.
    is_sub_agent: bool = False

    personality_traits: list[str]
    personality_quirks: list[str] | None
    voice_language: str
    voice_gender: str
    voice_pace: str

    character_id: str | None
    accent_color: str
    desk_position: int | None

    status: str
    mood: str
    current_task_id: uuid.UUID | None

    created_at: datetime
    updated_at: datetime

File: frontend/src/types/agent.ts
export type EngineType = "cli" | "api";

export type AgentStatus = "idle" | "working" | "thinking" | "break" | "offline";
export type AgentMood = "neutral" | "happy" | "stressed" | "excited" | "bored";

export interface Agent {
  id: string;
  name: string;
  role: string;

  engine_type: EngineType;
  engine_provider: string;
  engine_model: string | null;
  engine_command: string | null;
  working_directory: string | null;
  allowed_tools: string[] | null;
  // Never the raw key — just whether a bring-your-own API key is configured.
  has_engine_api_key: boolean;
  // Explicit opt-in for generate_image/generate_video tools — not derived
  // from engine_provider/key sharing (see backend media/registry.py).
  is_media_specialist: boolean;
  // Whether this agent is anyone's teammate (agent_collaborators) — drives
  // the "only chat with main agents" rule (see AgentCard.tsx).
  is_sub_agent: boolean;

  personality_traits: string[];
  personality_quirks: string[] | null;
  voice_language: string;
  voice_gender: string;
  voice_pace: string;

  character_id: string | null;
  accent_color: string;
  desk_position: number | null;

  status: AgentStatus;
  mood: AgentMood;
  current_task_id: string | null;

  created_at: string;
  updated_at: string;
}

export interface AgentCreate {
  name: string;
  role: string;
  engine_type: EngineType;
  engine_provider: string;
  engine_model?: string | null;
  engine_command?: string | null;
  working_directory?: string | null;
  allowed_tools?: string[] | null;
  // Plaintext here only — encrypted server-side, never returned. Required
  // for a custom (bring-your-own) API provider, i.e. engine_provider
  // outside {"anthropic", "ollama"}.
  engine_api_key?: string | null;
  is_media_specialist?: boolean;
  personality_traits: string[];
  personality_quirks?: string[] | null;
  voice_language?: string;
  voice_gender?: string;
  voice_pace?: string;
  character_id?: string | null;
  accent_color?: string;
  desk_position?: number | null;
}

export interface AgentUpdate {
  name?: string;
  role?: string;
  engine_type?: EngineType;
  engine_provider?: string;
  engine_model?: string | null;
  engine_command?: string | null;
  working_directory?: string | null;
  allowed_tools?: string[] | null;
  // Same contract as AgentCreate's field: omit to leave untouched, "" to
  // clear, a value to rotate. Never present in any response.
  engine_api_key?: string | null;
  is_media_specialist?: boolean;
  status?: AgentStatus;
  mood?: AgentMood;
  [key: string]: unknown;
}

export interface WorkspaceEntry {
  name: string;
  path: string;
  type: "file" | "dir";
  size: number | null;
}

export interface WorkspaceListing {
  path: string;
  entries: WorkspaceEntry[];
  host_path: string | null;
}

export interface WorkspaceFileContent {
  path: string;
  size: number;
  readable: boolean;
  content?: string | null;
  reason?: string | null;
}

export interface SoulUpdate {
  content: string;
}

export interface SoulRead {
  content: string;
}

export interface CollaboratorsUpdate {
  collaborator_ids: string[];
}

export interface CollaboratorsRead {
  collaborators: Agent[];
}
